File: Hard/number_of_digit_one/number_of_digit_one.py
from typing import List

def countDigitOne(n: int) -> int:
        count = 0
        factor = 1

        while factor <= n:
            lower = n % factor
            current = (n // factor) % 10
            higher = n // (factor * 10)

            if current == 0:
                count += higher * factor

            elif current == 1:
                count += higher * factor + lower + 1

            else:
                count += (higher + 1) * factor

            factor *= 10

        return count

if __name__ == "__main__":
    print("Testing countDigitOne function:")
    test_case_1 = 13
    print(f"For the input {test_case_1}, the output is:", countDigitOne(test_case_1)) #Output: 6

    test_case_2 = 0
    print(f"For the input {test_case_2}, the output is:", countDigitOne(test_case_2)) #Output: 0

File: Hard/number_of_digit_one/number_of_digit_one.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>
using namespace std;

class Solution {
    public:
        int countDigitOne(int n) {
            int count = 0;
            long factor = 1;

            while (factor <= n) {
                long lower = n % factor;
                long current = (n / factor) % 10;
                long higher = n / (factor * 10);

                if (current == 0) {
                    count += higher * factor;
                }
                else if (current == 1) {
                    count += higher * factor + lower + 1;
                }
                else {
                    count += (higher + 1) * factor;
                }

                factor *= 10;
            }

            return count;
    }
};

int main() {
    Solution solution;
    vector<int> testCases = {13, 0};
    for (int n : testCases) {
        int result = solution.countDigitOne(n);
        cout << "Number of digit '1' from 1 to " << n << ": " << result << endl;
    }
    return 0;
}